core/vm/analysis.go

```diff
@@ -19,6 +19,45 @@ package vm
 import (
 )
 
+// fill in segment of operation information array for a block
+func analyzeBlock(ctx *callCtx, pc uint64) (*BlockInfo, error) {
+	info := NewBlockInfo(ctx.contract, pc)
+	code := ctx.contract.Code
+	jumpTable := ctx.interpreter.jt
+	for ; pc < uint64(len(code)); pc++ {
+		op := OpCode(code[pc])
+		operation := jumpTable[op]
+		if operation == nil {
+			return nil, &ErrInvalidOpCode{opcode: op}
+		}
+		info.minStack += operation.minStack
+		info.maxStack += operation.maxStack
+		info.constantGas += operation.constantGas
+		
+		if op >= PUSH1 && op <= PUSH32 {
+
+			// decode push data to PushInfo
+			// ...			
+			continue
+		}
+		if op == JUMP || op == JUMPI {
+			prevOp := OpCode(code[pc-1])
+			if prevOp >= PUSH1 && prevOp <= PUSH32 {
+
+				// replace with JMP NOOP or JMPI NOOP and attach JumpInfo to JMP or JMPI
+				// ...
+			}
+			// end block
+			break
+		}
+		if 	op == JUMPDEST || op == STOP || op == RETURN || op == REVERT || op == SELFDESTRUCT {
+			// end block
+			break
+		}
```

Comment:
Must you not also break on `CALL` derivates?

Reply:
Calls inside the basic block are fine provided you have precise information about the gas usage and the call point. Same is needed for SSTORE and GAS instructions.

Reply:
`GAS`, yes, and basically anything where the dynamic portion of the gas is affected by the argument, no? So `SSTORE`, `SLOAD`, `EXTCODEHASH`, `EXTCODECOPY`, `BALANCE` etc which now has a gascost depending on whether the arg is hot or cold.

Reply:
No. These can fail earlier then when executing "normally", but execution would fail in this block anyway.
This is only for instruction which need exact gas usage: GAS by definition, SSTORE because of the `gas >= 2300` check, CALL because 63/64 rule and CREATE because it passes all remaining gas. Maybe there are some others which I don't recall now.

Reply:
So is this OK or not?

Reply:
I think it is OK. As far as I can see, the Blocks are only used to check stack underflow/overflow upfront, and also to consume `constantGas` upfront. So if it is detected upfont that there won't be enough gas for a block anyway, and it will OutOfGas, then there is no point in running all the nested calls, all the state changes will be reverted by Block going OutOfGas.